feat(lsp): support pull diagnostics on save via flag

Delerme opened the pull request:
Problem:
Some lsp servers can be configured to refresh diagnostics on save to avoid performance issues. The introduction of pull diagnostics in 0.10 goes around those settings by requesting diagnostics with every change event. See #29792

Solution:
Add an optional lsp client flag to request diagnostics on save if the lsp server supports pull diagnostics. By default, the existing behavior is maintained and diagnostics are requested with every change event, if the server supports pull diagnostics.

Commit:
feat(lsp): support pull diagnostics on save via flag

Problem:
Some lsp servers can be configured to refresh diagnostics on save to avoid
performance issues. The introduction of pull diagnostics in 0.10
goes around those settings by requesting diagnostics with every change
event. See #29792

Solution:
Add an optional lsp client flag to request diagnostics on save if the
lsp server supports pull diagnostics. By default, the existing behavior is
maintained and diagnostics are requested with every change event, if
the server supports pull diagnostics.

justinmk commented:
Does it really need to be optional? And, is it not possible to "configure" this via client/server capabilities?

Delerme commented:
> Does it really need to be optional?
I can make it required

> Is it not possible to "configure" this via client/server capabilities?
The user could override the capabilities to disable pull diagnostics entirely, but otherwise I'm not aware of something in the capabilities that would work. It seems like pull diagnostics are the "future" of lsp diagnostics so that the client is in control of when to pull, instead of every server having different implementations and settings. If so, it seems useful for neovim to start supporting settings for it. I could imagine this becoming a full field on vim.lsp.ClientConfig later, like `pull_diagnostics_on` with settings like `change`, `save`, `normal_mode`, etc. Open to other thoughts and ideas!

Delerme commented:
I may need some help with writing the functional test

I tried duplicating [this test](https://github.com/neovim/neovim/blob/a8fbe1d409e08c68b05bc26b096486020ae3162b/test/functional/plugin/lsp_spec.lua#L1295) and adjusted:
- added `diagnosticProvider` capability to enable pull diagnostics
- check for `textDocument/diagnostic` event sent to lsp sever

I added assertions checking that the capabilities were set correctly and that the [correct autocmd](https://github.com/neovim/neovim/blob/a8fbe1d409e08c68b05bc26b096486020ae3162b/runtime/lua/vim/lsp/inlay_hint.lua#L267) was on the buffer but I couldn't get it to fire. I even tried manually execing the autocmd in an `exec_lua` block but the callback didn't fire. Not sure if there's something special with autocmds in tests, or if you have any other advice!

mfussenegger commented:
> Some lsp servers can be configured to refresh diagnostics on save to avoid performance issues. The introduction of pull diagnostics in 0.10 goes around those settings by requesting diagnostics with every change event. See https://github.com/neovim/neovim/issues/29792

Couldn't you turn off pull in that case and rely on push?

> It seems like pull diagnostics are the "future" of lsp diagnostics so that the client is in control of when to pull, instead of every server having different implementations and settings. If so, it seems useful for neovim to start supporting settings for it. I could imagine this becoming a full field on vim.lsp.ClientConfig later, like pull_diagnostics_on with settings like change, save, normal_mode, etc. Open to other thoughts and ideas!

Afaik part of the rationale for a pull in addition to a push was that the client could have more control over which document to get diagnostics for first, to get them faster. It's not replacing push but complementing. 

Adding this new flag would imho go a bit against that and we definitely don't want to have a dozen settings to control when it should trigger.
Instead what we could do is expose enable and refresh and leave it up to the user to setup autocmds as they please. That would give maximum flexibility without having to introduce just one more option whenever someone can think of a new scenario. 

But I'm overall not convinced this is needed

Delerme commented:
> Couldn't you turn off pull in that case and rely on push?
> Afaik part of the rationale for a pull in addition to a push was that the client could have more control over which document to get diagnostics for first, to get them faster. It's not replacing push but complementing.

Ah that's helpful context, thanks!

> Adding this new flag would imho go a bit against that and we definitely don't want to have a dozen settings to control when it should trigger. Instead what we could do is expose enable and refresh and leave it up to the user to setup autocmds as they please. That would give maximum flexibility without having to introduce just one more option whenever someone can think of a new scenario.

Yes, given that info that makes more sense. What do you think about for now adding a simpler setting in the lsp client config to disable pull diagnostics? Right now it requires some relatively deep knowledge of lsp servers, capabilities, and how neovim implemented pull diagnostics. I'll admit when I tried to disable them I couldn't get it to work. It could be in `flags` or in the `ClientConfig` directly, but something like `pull_diagnostics boolean` or `disable_pull_diagnostics boolean`? What do you think?
